### ISSUE TITLE
Feature Request / Help Request

### ISSUE THREAD

**Issue by ColAngel** [bo3b/3Dmigoto#60]
I would add a feature request accorting to the posts #871, #872, and #1349 (3Dmigoto now open-source...).
Basically, it would be nice to have a feature that allows to force custom resolution, but not the way 3dmigoto does it now. The games should still operate on the resolution that is choosen within the game. Additionally the interlacing 3d method could be added to the custom shader (the upscaling and interlacing can be done whithin the custom shader i guess). This would have a nice effect for the people who uses passive 3d. I have tried Tridef with interlacing and upscaling to 4K and it looks much prettier than side by side. Even if the game is set to FullHD resoution and upscales to 4K it looks much better than 2K or even 4K side by side. And as a bonus you gain perfomance with very low quality loss.
I try to do it by myselft, but it will require some help.  I have already took a look at it and so far figured out following things:
 - The whole migoto work applies after the game done its stuff and before the actual present call is done. 
 - The present call is wrapped by the migoto and here is the place where the migoto magic happens, including the invokation of the custom shader. 
- The custom shader receives a texture that contains the output 2d image (one ivokation for left eye and one for the right).
-  Then the actual present call will be applied (hackerDXGI.cpp line 1853) 
What i plan to do is:
Initialization Step.
- create a swap chain render target with custom (forced) resolution and use it as render target.
- create CustomRenderTarget with games resolution and force it as the render target of the game (instead of the normal swap chain back buffer)
Per Frame Step:
- change the custom shader by adding interlacing and some kind of upscaling (very simple one for the beginning)
- set swap chain as back buffer for the custom shader invokation
Could it work this way? Maybe there is another better way to do it?  
Thank you in advance! :-)

**Comment by bo3b** [comment]
This seems to me like a workable approach.  You basically want to maintain two different swapchains, one at the high DSR type resolution, and the normal one for the game.

I'm not certain if this is possible, because the CreateSwapChain takes the ID3D11Device as an input, and I don't know offhand if you can have multiple swap chains per device.  

The other concern would be whether 3D Vision Automatic would know about the two swap chains and handle them properly.  You basically would need Automatic to handle both swap chains, because it would be required for the game itself, in order to draw in 3D and run the fixed shaders.  Then you'd also need the part where at Present, it takes a 2x width swapchain and alternately shows each eye.  I don't think anyone has ever tried that, so there's no way to know if that works.

Here is a probably related stackoverflow thread:  [http://stackoverflow.com/questions/10744788/display-different-images-per-monitor-directx-10?answertab=active#tab-top](url)

One device, but two swapchains.   

In terms of getting this to work, I'd recommend starting with just the crux of the problem, which is the dual swapchains, in some test program instead of in a full game.  Games tend to be freaky and complicate things.  

You could use the Tutorial07 for example, drop in 3Dmigoto and have it force full screen and Automatic mode.  If you can get this piece working, then the rest of it would definitely be doable.

**Comment by ColAngel** [comment]
Hi! Thank you for the reply!

Actually i dont want to create 2 swap chains. I would create a a texture and use it as the render target ( something like this http://www.rastertek.com/dx11tut22.html) Im not yet quite familiar with Direct X (opengl guy, and with OpenGL it would work this way:  something like creating custom frame buffer object and bind it in the right place). I would try to hack into OMSetRenderTargets where the game swap chain is bound and bind the custom texture instead. If it works that way the game and would know nothing about it neither the 3d vision automatic. Then in the hacked presend call would bind the actuall swap chain and then use the custom texture as in put in the custom shader, where the interpolation will be applied. 
The question is, is it actually possible to do it this way with directx in general (redirection the whole rendering from swap chain to the offscreen buffer) and how the 3d vision automatic will react to it.
Second question is where exactly the fixed shader are bound. (in the present call or in place via hacked shader bind interfaces).   

Your suggetion with too swap chains seems to be even better solution (if it works). 
I ll try both ways. Eitherway i need to make myself more familiar with directX, 3d vision automatic, and migoto. 

P.S.: the lastes posts on  https://forums.geforce.com/default/topic/685657/3d-vision/3dmigoto-now-open-source-/91/ suggests that it is allready working somehow  (via forcing custom resolution in the config file, but i can't imagine how) . Ill try to verify this tonight.

**Comment by DarkStarSword** [comment]
I actually did something similar to what you are after in WATCH_DOGS2 to work around a driver/game bug on non-SLI systems causing HUD flicker purely using the existing features in 3DMigoto to redirect the final render target to my own custom render target, then copying to the real back buffer on the present call - you could follow that as an example and use the width & height or width_multiply & height_multiply options in the [Resource] section to set your custom resolution on your replacement back buffer. In that case I matched any shader I knew to be used to render directly to the final render target, but you might alternatively consider adding some smarts to 3DMigoto to trigger a command list when a render target corresponding to the real back buffer is assigned to the pipeline. Notably, one gotcha I found was that the colours were wrong and I had to use the sRGB formula in the custom shader I used to perform that final copy to match the original colours from the game.

**Comment by DarkStarSword** [comment]
The above comment is somewhat edited FYI

**Comment by ColAngel** [comment]
Thank you for the help and suggetions :-) Yeah it is gonna be a fight :-) 

I have tried to compile migoto ( with VS 2015) and run into some issues. Some of the project internal headers cound not be found (had to add paths in affected projects) and it was necessary to comment out datatype redefinitions in binarydecompiler.h lines 620 - 625. They caused ambigios definition of the datatypes (I think it is something vc 14 compiler specific). I ve got it to work but maybe there is something Im missing at this point. 

@DarkStarSword 
Yeah i was planning to detect the binds of the swap chain and corresponding depth/tencil buffer (hack in to OMSetRenderTarget) and replace them with my own custom render target (heh, sounds much easier as it gonna be :-) ).  I want also find a general solution if it is possible at all :-). 
In general i would not commit anything to the git hub until im sure that it works and you or bo3b had took a look at it. Also a question: where i can find the Watch Docs 2 specific code? Is it in the main branch?

**Comment by DarkStarSword** [comment]
Both the decompiler and assembler are broken with vs2015 / win 8.1 SDK due to incompatible changes in d3dcompiler, so you need to use vs2013 / win 8 SDK (you can still use vs2015 for the IDE + better debugger if you like, so long as vs2013 and win 8 SDK is installed, or if you can manage without the internal decompiler and assembler (such as by using a prebuilt cmd_Decompiler in their place) you can use the work in progress forward port in the vs2015 topic branch - last time I checked the master branch still merged cleanly into it).

The WATCH_DOGS2 fix is in a separate tree:
https://github.com/DarkStarSword/3d-fixes-WATCH_DOGS2

**Comment by ColAngel** [comment]
Found both the vs 2015 branch and the watch dogs.  The vs 2015 version compiles without issues.
I ll try to debug into migoto and make myself more familiar with it and with the watch dog branch. It will cetrain take a while.
And again, thank you for your help. I will post an update as soon as i ll have made some progress with the feature.

**Comment by ColAngel** [comment]
As i promized the update:
Ich have experimented around with directX and Migoto and achived following.
I ve created an simple application bases on rastertek tutorial and done everything
i was planned to do in the example to check if my idea works at all (upsampling and custom rendertarget). It worked flawlessly.
Then i ve tried implement all this within migoto (hardcoded for now to keep everything simple) 
and it worked. (Used a very straith forward directx application that uses render to texture together with normal rendering (tutorial 22 rastertek without any changes).
What i have done:
1. Created the windows with a resolution (lets call it game resolution, as if it was set by the game config or something like that).
2. Then in the CreateSwapChain within migoto, i have changed the resolution 
	of the swap chain  to the custom  resolution (lets call it migoto resolution) and resized window by ResizeTarget function.
3. Actally created custom render target with game resolution. (tried to created second swap chain as a custom render target, it seems to work either)
4. In CreateRenderTargetView ive checked if the resource param == swapchain buffer and if yes created rendertargetview for the custom rendertarget
   and returned it to the game. At the same time i have created rendertarget view for the actuall swapchain and stored it within migoto.
5.6. In the Present function. I have inserted code to draw a quad that uses the custom render target as input and swap chain render target the render target.
(This causes some migoto functionality has been broken, but for sake of the simplicity it does not important at this time point.)  

I guess it is another way round of that what you suggested(I dont set resolution within game and downscale it within migoto, but let the game resolution untouched
 and use migoto to force desired resolution eg. game full hd migoto upscales to 4K). 
I thing this way we can keep things general and independend and only intercept on the end of the whole pipeline.

This method has one major draw back. The game window has to be resized and forced to keep the setting via migoto independently of the game. For some games
it is enough to use the already avalibale resizing function. Other games like witcher 3 seems to react differently(possible related to the posts i have already mentioned in comments bevore) .  If i call ResizeTarger the srceen mode changeds at first.
but it goes immediately back to game screen resolution (and somehow turns the fullscreen off). If i force the custom resolution within the resizeTarget then it flips
from one resolution to another. It seems the game react to the WM_SIZE caused by ResizeTarget (changing to the fullscreen) by using Win-API functionality to Resize Window and uses settings from the config).
And as you certainly know, there is a bunch of methods that can be used to resize window. Wenn need to hook them all or something like that. I guess this is also the way the TriDef
forces the resolution. Is there already similar functionality besized the hook of SetWindowPos?

And i have detected strange bug with debug version. It crashes somewhere in HackerContext::AfterDispatch(DispatchContext *context). Release version however works.

**Comment by ColAngel** [comment]
Hello Guys! 
I m done with upscaling (almost). Tested it with some games i have. 
I ve added fake swap chain in HackerSwapChain (texture and real second swap chain), Sampler - command for the custom shader and some other stuff to get it work all together. 
Independently of upscaling i have updated the 3dvisionSBS-pixel shader to support line interlacing (mode 6 ,7 ).  
To activate upscaling: In [device] section remove comments on width height and set up desired resolution.
add upscaling command (0 - deactivated, 1 - activated (dont use resize target), 2 - another way to activate it  (use resize target) and upscale_mode (0 - use texture2d and 1 - use second swap chain)
if upscaling is activated, "old" migoto resizing is deactivated. Additionally i ve added a new custom shader
[CustomShaderUpscale] that uses f_bb (fake bb) as input. It can be combined with custom shader for sbs
Tested it with: Witcher 3 (works), Batman Akrah Orignings (works), AssassingsCreed Black Flag (works), 
Darksiders 1 remastered (crashes i guess multithreading issue but dont know how and why yet) GTA V crashes, i m not sure yet but it  seems the game uses d3d11createdeviceandSwapchain- the function has no support for upscaling yet).  
I would commit it somewhere for you to take a look at it. ( ill provide filenames and code lines numbers i ve added and/or changed to make code review easier. ) 
Thanks again for support!

**Comment by bo3b** [comment]
That sounds pretty promising.  Great stuff!

Given that is crashing in a couple of test games, that suggests we should review this before adding it to 3Dmigoto.  We don't mind adding speculative features, but want to make sure core functionality is not affected.

For this, let's go ahead and use a pull request, then we can make concrete suggestions.

Here is a quick tutorial for creating a pull request:
[https://yangsu.github.io/pull-request-tutorial/](url)

Basically, you'll want to fork the 3Dmigoto master branch, make a new feature branch, and commit your changes to that branch. Then request that we take a look, and we'll get notified and can do direct comparisons.

If that all seems right, we'll go ahead and accept the pull request. Later on, if you might want to continue working on this, we can provide you with direct access to the 3Dmigoto repo.

Thanks for working on this.

**Comment by ColAngel** [comment]
Thanks for respond!
"...want to make sure core functionality is not affected" :  I have paid particular attention to keep everything the way it was, so that no core functionality is affected, (The changes in code are explicitely activated if the upscaling flag is on. But anyway, i would appreciate if somebody would review the code. 
And test it with more games! (I have not so many of them). 
By the way  i have found the cause of crash on darksiders remastered (The game does very weird stuff :-)):  the order of the last lines in CommandList.Cpp ResourceCopyOperation::run at the pretty end of the function (it is also the end of the file) seemed to cause all the trouble. I had to change it  from 
scr_resource->Release()
if (src_view)
     src_view->Release(),
to 
if (src_view)
     src_view->Release(),
if (src_resource)
     scr_resource->Release()

It seems cause of the fake swap chain it is somehow possible that if the scr_source is released before the view, the becomes invalid). I guess this change should not affect anything (other games). (I had no time to test it yet, will do it later today). Anyway the darsiders works flawessly. 
I try to commit everything today. (I used the Migoto for VS 2015 branch)

"...we can provide you with direct access to the 3Dmigoto repo" that would be nice. If Im on the right path, then i would definitely keep working on that (or maybe on other stuff later).

And you dont have to thank me, Im using the fixes and migoto since it is available! Me and all other 3d vision users have to thank you all :-)

**Comment by llyzski** [comment]
@ColAngel 

I have written a wiki for this feature: https://github.com/bo3b/3Dmigoto/wiki/Upscale-resolution

Please edit if you find anything inaccurate. Thanks for providing the feature!

**Comment by ColAngel** [comment]
@llyzski, Thank you for writting the wiki for the feature.  I have no time at the moment to take a deeper look at it but it seems ok!
One request to you, please dont close issues you are not directly working on, because you can not know if everything was planned to do is done. In this special case it does not matter, because i was planning close it anyway. 
I want to open another issue later that handles follow situation:
currently the feature works only with swap chains that are created with Factory::createSwapChain(...) method and it seems ok for the moment, due to the fact i was not able to find a game that uses another method. (There are 4 or 5 other possibilities to create swap chain that need to be updated to support upscaling).  There is also a DXGIFactory1 that need to be handled. 

Thank you in advance!

**Comment by llyzski** [comment]
Ah OK, I though it is generally OK to close an issue on github because you can reopen an issue any time, but I will be careful next time. In fact I was following your development for quite some time, using this feature myself and tested it quite a lot, so thanks for developing it!
